[ZIP 321] Introduce Asset ID to payment url (#1101)
zcash/zips: zips/zip-0321.rst

Reviewer: ```suggestion

The term "Custom Asset" in this document is to be interpreted as described
in ZIP 227 [^zip-0227].

```

==========

In order for a robust transactional ecosystem to evolve for Zcash, it is
necessary for vendors to be able to issue requests for payment. At present,
the best option available is to manually specify a payment address, a payment
amount, and potentially asset id and memo field content. Of these three
components, existing wallets only provide functionality for reading payment
addresses in a semi-automated fashion. It is then necessary for the user to
manually enter payment asset ids, amounts and any associated memo information,
which is tedious and may be error-prone, particularly if a payment is intended
for multiple recipients or the memo field information contains structured data

> that must be faithfully reproduced.

Reviewer 2: This change to the motivation doesn't make sense, because asset IDs did not exist when the ZIP was written (and the "Of these three" components becomes incorrect after the edit). Instead, change "At present" to "At time of first writing this ZIP".

The format must be a valid URI format [#RFC3986]_.

> The format must permit the representation of one or more (payment address,
> asset id, amount, memo) tuples.

Reviewer: ```suggestion
The format must permit the representation of one or more tuples of
payment address, amount, asset id (if applicable), and optional memo
tuples.
```

  zcashurn        = "zcash:" ( zcashaddress [ "?" zcashparams ] / "?" zcashparams )
  zcashaddress    = 1*( ALPHA / DIGIT )
  zcashparams     = zcashparam [ "&" zcashparams ]
  zcashparam      = [ addrparam / amountparam / assetidparam / memoparam / messageparam / labelparam / reqparam / otherparam ]
  NONZERO         = %x31-39
  DIGIT           = %x30-39
  paramindex      = "." NONZERO 0*3DIGIT
  addrparam       = "address" [ paramindex ] "=" zcashaddress
  amountparam     = "amount"  [ paramindex ] "=" 1*DIGIT [ "." 1*8DIGIT ]

>   assetidparam    = "asset_id"[ paramindex ] "=" 88*88base64url

Reviewer 3: Given the constraints of the `Forward Compatibility` section, this must be a `req-` parameter.

```suggestion
  assetidparam    = "req-asset_id"[ paramindex ] "=" 88*88base64url
```

Reviewer: `_` is not allowed.

that use ``qchar``, i.e. the values of label, message, ``reqparam``, and
``otherparam`` parameters.

> The value of an ``asset_id`` parameter MUST be a base64url string without

Reviewer 3: ```suggestion
The value of an ``req-asset_id`` parameter MUST be a base64url string without
```

that use ``qchar``, i.e. the values of label, message, ``reqparam``, and
``otherparam`` parameters.

The value of an ``req-asset`` parameter MUST be a `base64url` string without
padding. This string is obtained by concatenating a version byte, the
canonical ``EncodeAssetId`` defined in ZIP 227 [#zip-0227]_, and an asset

> amount encoded as an unsigned integer in minimal units.

Reviewer 4: ```suggestion
amount encoded as an unsigned integer value representing the number of Asset tokens,
in the smallest possible denomination for the Asset.
```

applies to other address types in future, consideration should be given to
updating this ZIP to exclude these types, as part of their deprecation.

Asset Identifiers
-----------------

If an asset identifier is provided, it MUST be specified as the string returned

> by ``EncodeAssetId`` from ZIP 227 [#zip-0227]_. This value is defined as:
> 
> ::
>   EncodeAssetId = 0x00 || issuer || assetDescHash
> 
> where:
> 
> * ``issuer`` is ``0x00 || ik``, with ik a 32-byte public key, and
> * ``assetDescHash`` is the 32-byte BLAKE2b-256 digest of the asset description,
>   using personalization "ZSA-AssetDescCRH".
> 
> This encoding yields 66 bytes, which MUST be represented in a URI using base64url
> without padding. Implementations MUST decode this string back to exactly 66 bytes.

Reviewer: ```suggestion
by ``EncodeAssetId`` as specified in ZIP 227 [#zip-0227]_.
```

applies to other address types in future, consideration should be given to
updating this ZIP to exclude these types, as part of their deprecation.

Assets
------

The ``req-asset`` parameter permits the specification of both an asset
identifier and an amount for a payment. It is encoded as follows:

::

>   req-asset = base64url( 0x00 || EncodeAssetId(66 bytes) || uint64_be(8 bytes) )

Reviewer 4: ```suggestion
$\textt{req-assert} = \mathsf{base6url}\big(\mathtt{0x00} \,||\, \mathsf{I2LEOSP}_{64}(value) \,||\, \mathsf{EncodeAssetId}(\mathsf{AssetId})\big)$
```

The leading version byte (``0x00``) indicates the encoding format.
For version ``0x00``, the next 66 bytes are the ``EncodeAssetId`` value
defined in ZIP 227 [#zip-0227]_, followed by 8 bytes representing the amount
as an unsigned 64-bit integer in big-endian (network) byte order,

> corresponding to the number of minimal units of the asset to be transferred.

Reviewer 4: ```suggestion
The leading version byte (``0x00``) indicates the encoding format.
For version ``0x00``, the next 8 bytes represent the amount
as an unsigned 64-bit integer in little-endian byte order,
followed by the encoded $\mathsf{AssetId}$.
```

the remainder of the parameter.

If no ``req-asset`` parameter is present, the payment MUST be considered to
be for the default asset (ZEC), and the transfer amount is specified using
the ``amount`` parameter.

Only Orchard receivers are valid for asset payments. Therefore, a payment request
MUST use a Unified Address that encodes at least one Orchard receiver.
If the address in a payment request is a transparent address or a Sapling address,
then the request is invalid.

Example:

Suppose the asset description is "USD testing", with an invented digest and
issuer key as inputs. The resulting 66-byte value is:

::

  0000deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
  11223344556677889900aabbccddeeff00112233445566778899aabbccddeeff

We want to create a URI to request a payment of ``100.55`` of this asset:

::

  100.55 USD = 10055 minimal units (assuming 2 decimal places)
  = 0x0000000000002747
  = 00 00 00 00 00 00 27 47 (8 bytes, big-endian)

Concatenating the version byte, asset id, and amount gives (hex):
::

  00
  0000deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
  11223344556677889900aabbccddeeff00112233445566778899aabbccddeeff

>   0000000000002747

Reviewer 4: ```suggestion
  = 47 27 00 00 00 00 00 00 (8 bytes, little-endian)

Concatenating the version byte, amount, and asset id gives (hex):
::

  00
  4727000000000000
  0000deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
  11223344556677889900aabbccddeeff00112233445566778899aabbccddeeff
```

Base64url-encoded (unpadded) for the entire 75-byte blob, this becomes
(100 characters):

::

>   AAAA3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-796tvu8RIjNEVWZ3iJkAqrvM3e7_ABEiM0RVZneImaq7zN3u_wAAAAAAACdH

Reviewer 4: Please re-encode based on the schema changes above

A payment URI using this ``req-asset`` could look like:

::

  zcash:utest1q9d6gqg...7y0d9zj7ns2ceqk4j0d6ngqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7e4t0p

>     ?req-asset=AAAA3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-796tvu8RIjNEVWZ3iJkAqrvM3e7_ABEiM0RVZneImaq7zN3u_wAAAAAAACdH

Reviewer 4: Please re-encode as above

Transfer amount
---------------

> If an amount is provided without an asset parameter, the amount is considered to be

Reviewer 4: ```suggestion
If an amount is provided without a ``req-asset`` parameter, the amount is considered to be
```

updating this ZIP to exclude these types, as part of their deprecation.

Assets
------

The ``req-asset`` parameter permits the specification of both an asset
identifier and an amount for a payment. It is encoded as follows:

$\mathtt{asset} = \mathsf{base64url}\big(\mathtt{0x00} \,||\, \mathsf{I2LEOSP}_{64}(value) \,||\, \mathsf{EncodeAssetId}(\mathsf{AssetId})\big)$

The leading version byte (``0x00``) indicates the encoding format.
For version ``0x00``, the next 8 bytes represent the amount
as an unsigned 64-bit integer in little-endian byte order,
followed by the encoded $\mathsf{AssetId}$.

Future versions MAY define different structures or lengths after the version
byte; implementations MUST use the version value to determine how to decode
the remainder of the parameter.

If no ``req-asset`` parameter is present, the payment MUST be considered to
be for ZEC, and the transfer amount in ZEC is specified using the ``amount``
parameter.

Only Orchard receivers are valid for Custom Asset payments. Therefore, a payment
request for a Custom Asset MUST use a Unified Address that encodes at least one
Orchard receiver.

> Example:

Reviewer 2: This cannot be part of the specification, and so if left inline it should be handled the same way as inline Rationale sections:

```suggestion
Example
~~~~~~~

.. raw:: html

   <details>
   <summary>Click to show/hide</summary>

```

Suppose the asset description is "USD testing", with an invented digest and
issuer key as inputs. The resulting 66-byte value is:

::

  0000deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
  11223344556677889900aabbccddeeff00112233445566778899aabbccddeeff

> We want to create a URI to request a payment of ``100.55`` of this asset:

Reviewer: ```suggestion
We want to create a URI to request a payment of ``10055`` minimal units of
this asset (hypothetically representing USD 100.55):
```

::

  100.55 USD = 10055 minimal units (assuming 2 decimal places)
  = 0x0000000000002747
  = 47 27 00 00 00 00 00 00 (8 bytes, little-endian)

Concatenating the version byte, amount, and asset id gives (hex):

::

  00
  4727000000000000
  0000deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
  11223344556677889900aabbccddeeff00112233445566778899aabbccddeeff

Base64url-encoded (unpadded) for the entire 75-byte blob, this becomes
(100 characters):

::

  AEcnAAAAAAAAAADerb7v3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-7xEiM0RVZneImQCqu8zd7v8AESIzRFVmd4iZqrvM3e7_

A payment URI using this ``req-asset`` could look like:

::

  zcash:utest1q9d6gqg...7y0d9zj7ns2ceqk4j0d6ngqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7e4t0p
    ?req-asset=AEcnAAAAAAAAAADerb7v3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-7xEiM0RVZneImQCqu8zd7v8AESIzRFVmd4iZqrvM3e7_

> 

Reviewer 2: ```suggestion

.. raw:: html

   </details>
```

Transfer amount

> ---------------

Reviewer: ```suggestion
ZEC Transfer Amount
-------------------
```

The ``req-asset`` parameter permits the specification of both an asset
identifier and an amount for a payment. It is encoded as follows:

> $\mathtt{asset} = \mathsf{base6url}\big(\mathtt{0x00} \,||\, \mathsf{I2LEOSP}_{64}(value) \,||\, \mathsf{EncodeAssetId}(\mathsf{AssetId})\big)$

Author: Thanks :sweat_smile: 

Fixed at https://github.com/zcash/zips/pull/1101/commits/9b69ea3e49cb1dc33214dad2776fab6188a88ef4

> The amount MUST NOT be greater than 21000000 ZEC (in general, monetary amounts

Reviewer 3: If this rule is enforced at the parser level at present, it means that parsers that add support for `req-assetId` must be cautious to also relax this parsing restriction.

Reviewer 3: In ZIP Editing, we identified that an additional `req-assetamount` parameter should be added. This should be base64 encoded, and should specify the value in the minimal unit. A URI that has both `amount` and `req-assetamount` should be rejected.

For assets other than ZEC, we have no way of determining how many decimal places there are.

An alternative suggested by @daira is that instead of two fields, a base64-encoded `req-asset` that encodes both the asset ID and the value should be used, because in the context of an opaque asset ID, the user has no frame of reference for interpreting the value of a `req-assetamount` field.

in Zcash cannot be greater than this value).

The same syntax and validity rules apply when specifying amounts for other assets;
in that case, the asset being transferred is identified by the accompanying

> ``asset_id`` parameter.

Reviewer 3: ```suggestion
``req-assetid`` parameter.
```

And similarly update the test vectors

A valid payment request with one transparent and one shielded Sapling recipient
address, with a base64url-encoded Unicode memo for the shielded recipient.

::

  zcash:?address=utest1qpsw0d7s6e5t7dv8x8yyw5e4v4jh8p5kux6g5v4kthh0nngwh2c5ue2x8p3wxy7qsqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7egr7p
    &req-asset=AEcnAAAAAAAAAADerb7v3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-7xEiM0RVZneImQCqu8zd7v8AESIzRFVmd4iZqrvM3e7_
    &address.1=utest1qplk9cnv5u2r0z8a3g7yyjv4x3v5h4d0nx3uq36zj2u4p6h9ty4g8r9xk2q0d2n7qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq4u3aq
    &req-asset.1=AJAIAwAAAAAAADerb7v3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-7xEiM0RVZneImQCqu8zd7v8AESIzRFVmd4iZqrvM3e7_

This is a valid payment request with two recipients, each using a distinct Unified
Address that includes an Orchard receiver. Each ``req-asset`` value encodes, in order,
a version byte (``0x00``), an 8-byte unsigned integer amount in minimal units, and
a 66-byte ``EncodeAssetId``.

> In this example both payments reference the same asset: the first requests 10 055

Reviewer: ```suggestion

In this example both payments reference the same asset: the first requests 10 055
```

A valid payment request with one transparent and one shielded Sapling recipient
address, with a base64url-encoded Unicode memo for the shielded recipient.

::

  zcash:?address=utest1qpsw0d7s6e5t7dv8x8yyw5e4v4jh8p5kux6g5v4kthh0nngwh2c5ue2x8p3wxy7qsqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq7egr7p
  &req-asset=AAAB3q2-796tvu_erb7v3q2-796tvu_erb7v3q2-796tvu8RIjNEVWZ3iJkAqrvM3e7_ABEiM0RVZneImaq7zN3u_wAAAAAAABkF
  &address.1=utest1qplk9cnv5u2r0z8a3g7yyjv4x3v5h4d0nx3uq36zj2u4p6h9ty4g8r9xk2q0d2n7qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq4u3aq
  &req-asset.1=AAAC2r1_795qsu9grb6u3q2-796tvu_erb7v3q2-796tvu8RIjNEVWZ3iJkAqrvM3e7_ABEiM0RVZneImaq7zN3u_wAAAAAAADhP

This represents a valid payment request with two recipients, each having a distinct
Unified Address that includes an Orchard receiver. Each ``req-asset`` parameter

> encodes a version byte, an asset identifier and asset amount.

Reviewer 4: ```suggestion
encodes a version byte, an asset amount, and an asset identifier.
```

"hierarchical URIs" in the sense defined in [#RFC3986]_ section 1.2.3,
and do not have an "authority component".

::

  zcash:?address=ztestsapling10yy2ex5dcqkclhc7z7yrnjq2z6feyjad56ptwlfgmy77dmaqqrl9gyhprdx59qgmsnyfska2kez
    &amount=5
    &asset_id=not@valid!chars

Invalid:
 * A Sapling address is used with an ``asset_id`` parameter. Sapling receivers do not
   support asset identifiers.
 * The ``asset_id`` contains characters (e.g. ``@``, ``!``) that are not permitted in
   base64url.

>  * The ``asset_id`` value is not exactly 88 characters long, and therefore cannot
>    decode to the required 66-byte ``EncodeAssetId``.

Reviewer 2: ```suggestion
```
